### PR TITLE
Add before/after strategies

### PR DESCRIPTION
**Pull request by harlowja**
Make these before and after functions more official
by providing some basic ones and a no-op ones that
will be used by default.

### CODE REVIEW

File: tenacity/after.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from tenacity import _utils
+
+
+def nothing_after(func, trial_number):
```
Comment (jd) [inline tenacity/after.py]: All functions so far are named `foo_bar()` from module `foo`, but here you used `bar_foo`. I know English is better read this way, but for the sake of consistency… ?

Comment (harlowja) [inline tenacity/after.py]: Sure sure will change :-P

**Comment by harlowja** [comment]
Ok should be good to go...

File: tenacity/__init__.py [hunk 2]
```diff
@@ -122,15 +130,20 @@ def call(self, fn, *args, **kwargs):
...
+            trial_start_time = now()
+            if self.before is not None:
+                self.before(fn, attempt_number)
 
             fut = Future(attempt_number)
             try:
-                fut.set_result(fn(*args, **kwargs))
+                result = fn(*args, **kwargs)
+                trial_end_time = now()
+                fut.set_result(result)
```
Comment (jd) [inline tenacity/__init__.py]: I think both L140 and L141 should be put in the `else` statement. I was already eager to do it last time, but the more I see thing in this `try` code that shouldn't raise `TryAgain` nor raise any `Exception` embedded, the more I am unhappy.

e.g. I don't want now() failing being a cause of "should I retry?" :-/

**Comment by jd** [comment]
Minor thing that I missed but I'm ok to fix it after :)

File: tenacity/__init__.py (hunk 2 above)
Comment (harlowja) [inline tenacity/__init__.py]: Works for me.

**Comment by harlowja** [comment]
Done.

**Pull request merged**